notebooks/optuna-example.ipynb

```diff
@@ -0,0 +1,495 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## constants"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "np.random.seed(0)\n",
+    "torch.manual_seed(0)\n",
+    "torch.cuda.manual_seed(0)\n",
+    "torch.cuda.manual_seed_all(0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
+    "METRIC_NAME = 'valid_acc'\n",
+    "DIRECTION = 'maximize'  # we want to optimize accuracy, hence higher values are better\n",
+    "SAMPLER = optuna.samplers.TPESampler(seed=0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## custom functions"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_data():\n",
+    "    X, y = make_classification(1000, n_informative=15, random_state=0)\n",
+    "    X = X.astype(np.float32)\n",
+    "    y = y.astype(np.int64)\n",
+    "    return X, y"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MyModule(nn.Module):\n",
+    "    def __init__(self, num_units=10, nonlin=nn.ReLU(), dropout_rate=0.5):\n",
+    "        super(MyModule, self).__init__()\n",
+    "        self.num_units = num_units\n",
+    "        self.nonlin = nonlin\n",
+    "        self.dropout_rate = dropout_rate\n",
+    "        \n",
+    "        self.reset_weights()\n",
+    "        \n",
+    "    def reset_weights(self):\n",
+    "        self.dense0 = nn.Linear(20, self.num_units)\n",
+    "        self.dropout = nn.Dropout(self.dropout_rate)\n",
+    "        self.dense1 = nn.Linear(self.num_units, self.num_units)\n",
+    "        self.output = nn.Linear(self.num_units, 2)\n",
+    "        self.softmax = nn.Softmax(dim=-1)\n",
+    "\n",
+    "    def forward(self, X, **kwargs):\n",
+    "        X = self.nonlin(self.dense0(X))\n",
+    "        X = self.dropout(X)\n",
+    "        X = self.nonlin(self.dense1(X))\n",
+    "        X = self.softmax(self.output(X))\n",
+    "        return X"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def define_model(trial):\n",
+    "    \"\"\"Model definition and hyper-parameter setting\n",
+    "\n",
+    "    We set the number of units, the optimizer, the learning rate, and the number of epochs.\n",
+    "    \n",
+    "    To make this work efficiently with optuna, add the\n",
+    "    `SkorchPruningCallback <https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html?highlight=optuna.integration.OptunaSearchCV>`\n",
+    "    from the optuna library.\n",
+    "\n",
+    "    \"\"\"\n",
```

Comment:
Given this is a notebook, can we pull this out into an markdown block above this code block?

Reply:
done